File: python/src/threads/evals/answers.py
"""The tool bodies of a rerun: every pinned tool answers from the case, never from a sandbox. A
saved case's read-only calls answer from sandbox.json v2 by (tool, args_hash, occurrence); the
corpus's scripted sandbox (the v1 shape) answers by tool name, with provider dedup, lookups and
process checks; mediated calls answer from the recorded stubs. A read-only call the recording
never made fails closed."""

from collections import Counter
from collections.abc import Callable, Mapping, Sequence
from typing import Final

from pydantic import JsonValue
from pydantic.experimental.missing_sentinel import MISSING

from threads._generated.eval_v1 import RecallRecord, SandboxResult
from threads._json_schema import holds
from threads.evals.case_dir import CaseSandbox
from threads.log import JsonObject, ParseError, ToolSpec
from threads.loop.model import (
    Found,
    LookupResult,
    LookupUnknown,
    NotFound,
    NotFoundNonfinal,
)
from threads.loop.stubs import StubGateway
from threads.loop.tools import Dispatched, Invocation, Output, Reference, Termination
from threads.result import Err, Ok
from threads.thread.case_files import args_hash

_RECALLS: Final[Mapping[str, str]] = {"search_memory": "memory", "search_knowledge": "knowledge"}
"""The recalling tools: their recorded items come back as the call's references."""

type ReadArtifact = Callable[[str], Ok[bytes] | Err[ParseError]]


def _references(record: RecallRecord) -> tuple[Reference, ...]:
    out: list[Reference] = []
    for item in record.items:
        source = item.source
        if source not in ("memory", "knowledge") or not isinstance(item.text, str):
            continue
        origin = item.origin
        version = origin.version if isinstance(origin.version, str) else ""
        location = origin.location if isinstance(origin.location, str) else None
        out.append(Reference(source, origin.id, version, item.text, location))
    return tuple(out)


def _text(value: JsonValue) -> str:
    return value if isinstance(value, str) else ""


class Answers:
    """The rerun's tool runner (loop/tools.py `ToolRunner`)."""

    def __init__(
        self,
        sandbox: CaseSandbox | None,
        recall: Sequence[RecallRecord],
        stubs: StubGateway | None,
        read: ReadArtifact,
        now: Callable[[], int],
    ) -> None:
        self._results = () if sandbox is None else sandbox.results
        self._v1: Mapping[str, Mapping[str, JsonValue]] = (
            {} if sandbox is None or sandbox.v1 is None else sandbox.v1
        )
        self._recall = list(recall)
        self._stubs = stubs
        self._read = read
        self._now = now
        self._seen: Counter[tuple[str, str]] = Counter()
        self._used: set[int] = set()
        self._recalled = 0
        self.unrecorded: int = 0
        self.dispatches: Counter[str] = Counter()
        self.new_executions: Counter[str] = Counter()
        self.lookups: Counter[str] = Counter()

    def invalid(self, spec: ToolSpec, input: JsonObject) -> str | None:
        """The pinned schema, checked with the log's own keyword subset; a keyword it can't
        check is taken as recorded (the recording's tool already accepted the call)."""
        full = self._full_spec(spec)
        if full.input_schema is MISSING:
            return None
        try:
            return None if holds(dict(full.input_schema), dict(input)) else "invalid input"
        except TypeError:
            return None

    def _full_spec(self, spec: ToolSpec) -> ToolSpec:
        """A deferred tool's full spec, from the artifact its reference form names."""
        if spec.spec_ref is MISSING:
            return spec
        got = self._read(spec.spec_ref.sha256)
        if isinstance(got, Err):
            raise AssertionError(f"a pinned spec_ref reads back: {got.error.message}")
        return ToolSpec.model_validate_json(got.value)

    async def dispatch(self, call: Invocation) -> Dispatched:
        name = call.spec.name
        if name in self._v1:
            return self._scripted(name, call.effect_key)
        if call.spec.effect_class != "read_only" and self._stubs is not None:
            return await self._stubs.dispatch(call)
        return self._recorded(name, call.input)

    def _scripted(self, name: str, effect_key: str) -> Output:
        tool = self._v1[name]
        self.dispatches[name] += 1
        keys = tool.get("executed_keys")
        done = keys.get(effect_key) if isinstance(keys, dict) else None
        if isinstance(done, str):
            return Output(done)
        self.new_executions[name] += 1
        return Output(_text(tool.get("output")), tool.get("is_error") is True)

    def _recorded(self, name: str, input: JsonObject) -> Output:
        key = (name, args_hash(input))
        self._seen[key] += 1
        found = next(
            (
                (i, r)
                for i, r in enumerate(self._results)
                if (r.tool, r.args_hash, r.occurrence) == (*key, self._seen[key])
            ),
            None,
        )
        if found is None:
            self.unrecorded += 1
            return Output(
                f"unrecorded_call: the saved turn made no {name} call with these arguments", True
            )
        index, r = found
        self._used.add(index)
        content = () if r.content is MISSING else tuple(r.content)
        refs = self._recalled_by(name)
        return Output(self._full(r), r.is_error, content=content, references=refs)

    def _full(self, r: SandboxResult) -> str:
        if r.ref is MISSING:
            return r.preview
        got = self._read(r.ref.sha256)
        return got.value.decode("utf-8") if isinstance(got, Ok) else r.preview

    def _recalled_by(self, name: str) -> tuple[Reference, ...]:
        source = _RECALLS.get(name)
        if source is None:
            return ()
        index = next((i for i, r in enumerate(self._recall) if r.source == source), None)
        if index is None:
            return ()
        self._recalled += 1
        return _references(self._recall.pop(index))

    async def lookup(self, call: Invocation) -> LookupResult[str]:
        name = call.spec.name
        tool = self._v1.get(name)
        if tool is None:
            return LookupUnknown(f"no recorded lookup for {name}")
        self.lookups[name] += 1
        answers = tool.get("lookup")
        answer = answers.get(call.effect_key) if isinstance(answers, dict) else None
        if not isinstance(answer, dict):
            return LookupUnknown("no scripted answer")
        final = answer.get("final") is True
        if answer.get("result") == "found" and final:
            return Found(_text(answer.get("output", tool.get("output"))))
        if answer.get("result") == "not_found":
            return NotFound() if final else NotFoundNonfinal()
        return LookupUnknown("not final")

    async def terminate(self, call: Invocation) -> Termination:
        tool = self._v1.get(call.spec.name, {})
        return "terminated" if tool.get("process") == "terminated" else "unknown"

    def provider_now(self) -> int | None:
        return self._now()

    def counters(self) -> dict[str, dict[str, int]]:
        return {
            "dispatches": dict(self.dispatches),
            "new_executions": dict(self.new_executions),
            "lookups": dict(self.lookups),
        }

    def left(self) -> int:
        """Recorded results and recall no call used."""
        return len(self._results) - len(self._used) + len(self._recall)

File: typescript/packages/core/src/evals/answers.ts
import { z } from "zod";
import type { ToolSpec } from "../log";
import type { ToolImpl, ToolRun } from "../loop";
import type { LookupResult } from "../model";
import type { SandboxScript } from "../sandbox/script";
import type { ArtifactStore } from "../store";
import { argsHash } from "../thread/case-files";
import type { RecallRecord, SandboxResult, SandboxResults } from "./files";

// The tool bodies of a rerun: every pinned tool answers from the case, never from a sandbox.
// A saved case's read-only calls answer from sandbox.json v2 by (tool, args_hash, occurrence);
// the corpus's scripted sandbox (the v1 shape) answers by tool name, with provider dedup,
// lookups and process checks. A read-only call the recording never made fails closed.

export type Counters = {
  readonly dispatches: Readonly<Record<string, number>>;
  readonly new_executions: Readonly<Record<string, number>>;
  readonly lookups: Readonly<Record<string, number>>;
};

export type Answers = {
  readonly tools: ReadonlyMap<string, ToolImpl>;
  readonly counters: () => Counters;
  /** Read-only calls with no recorded result. */
  readonly unrecorded: () => number;
  /** Recorded results and recall no call used. */
  readonly left: () => number;
};

/** The recalling tools: their recorded items come back as the call's injections. */
const RECALLS: Readonly<Record<string, RecallRecord["source"]>> = {
  search_memory: "memory",
  search_knowledge: "knowledge",
};

/** A pinned schema compiled for the rerun; anything it can't compile is taken as recorded. */
function inputOf(spec: ToolSpec): z.ZodType {
  if (spec.input_schema === undefined) return z.json();
  try {
    return z.fromJSONSchema(spec.input_schema);
  } catch {
    return z.json();
  }
}

type Counter = { [tool: string]: number };
type Input = Parameters<ToolImpl["run"]>[0];

function bump(map: Counter, tool: string): void {
  map[tool] = (map[tool] ?? 0) + 1;
}

/** v2: the recorded result of this call, found by its key; its full bytes when spilled. */
function recorded(
  results: readonly SandboxResult[],
  used: Set<SandboxResult>,
  artifacts: Pick<ArtifactStore, "get">,
) {
  const seen = new Map<string, number>();
  return async (tool: string, input: Input): Promise<ToolRun | undefined> => {
    const hash = argsHash(input);
    const key = `${tool}\n${hash}`;
    const occurrence = (seen.get(key) ?? 0) + 1;
    seen.set(key, occurrence);
    const r = results.find(
      (x) =>
        x.tool === tool && x.args_hash === hash && x.occurrence === occurrence,
    );
    if (r === undefined) return undefined;
    used.add(r);
    const full =
      r.ref === undefined ? undefined : await artifacts.get(r.ref.sha256);
    return {
      kind: "done",
      output:
        full?.ok === true ? new TextDecoder().decode(full.value) : r.preview,
      isError: r.is_error,
      ...(r.content === undefined ? {} : { content: r.content }),
    };
  };
}

type Scripted = NonNullable<SandboxScript["tools"]>[string];

/** The corpus's scripted sandbox for one tool: dedup, lookup and process answers. */
function scripted(
  spec: ToolSpec,
  t: Scripted,
  counters: { dispatches: Counter; new_executions: Counter; lookups: Counter },
  now: () => number,
): Omit<ToolImpl, "spec" | "input"> {
  return {
    run: async (_input, ctx) => {
      bump(counters.dispatches, spec.name);
      const deduped = t.executed_keys?.[ctx.effectKey];
      if (deduped !== undefined)
        return { kind: "done", output: deduped, isError: false };
      bump(counters.new_executions, spec.name);
      return { kind: "done", output: t.output, isError: t.is_error ?? false };
    },
    reconcile: {
      finality: "final",
      lookup: async (key): Promise<LookupResult<string>> => {
        bump(counters.lookups, spec.name);
        const answer = t.lookup?.[key];
        if (answer === undefined)
          return { status: "unknown", reason: "no answer" };
        if (answer.result === "not_found")
          return { status: answer.final ? "not_found" : "not_found_nonfinal" };
        return answer.final
          ? { status: "found", value: answer.output ?? t.output }
          : { status: "unknown", reason: "found without finality" };
      },
    },
    terminate: async () =>
      t.process === "terminated" ? "terminated" : "unknown",
    providerNow: now,
  };
}

export type AnswerSource = {
  readonly specs: readonly ToolSpec[];
  /** sandbox.json: v2 results, the corpus's scripted sandbox, or nothing. */
  readonly sandbox: SandboxResults | SandboxScript | undefined;
  readonly recall: readonly RecallRecord[];
  readonly artifacts: Pick<ArtifactStore, "get">;
  readonly now: () => number;
};

export function answers(source: AnswerSource): Answers {
  const { sandbox, recall } = source;
  const results =
    sandbox !== undefined && "results" in sandbox ? sandbox.results : [];
  const tools =
    sandbox !== undefined && "results" in sandbox ? {} : (sandbox?.tools ?? {});
  const counters = { dispatches: {}, new_executions: {}, lookups: {} };
  const used = new Set<SandboxResult>();
  const recalled = new Set<RecallRecord>();
  let unrecorded = 0;
  const lookup = recorded(results, used, source.artifacts);
  const nextRecall = (tool: string) => {
    const found = recall.find(
      (r) => r.source === RECALLS[tool] && !recalled.has(r),
    );
    if (found !== undefined) recalled.add(found);
    return found?.items;
  };
  const impls = new Map<string, ToolImpl>();
  for (const spec of source.specs) {
    const t = tools[spec.name];
    const base = { spec, input: inputOf(spec) };
    if (t !== undefined) {
      impls.set(spec.name, {
        ...base,
        ...scripted(spec, t, counters, source.now),
        ...(t.concurrent === true ? { concurrent: true as const } : {}),
      });
      continue;
    }
    impls.set(spec.name, {
      ...base,
      run: async (input) => {
        const run = await lookup(spec.name, input);
        if (run === undefined) {
          unrecorded += 1;
          return {
            kind: "done",
            output: `unrecorded_call: the saved turn made no ${spec.name} call with these arguments`,
            isError: true,
          };
        }
        const inject = run.kind === "done" ? nextRecall(spec.name) : undefined;
        return run.kind === "done" && inject !== undefined
          ? { ...run, inject }
          : run;
      },
    });
  }
  return {
    tools: impls,
    counters: () => counters,
    unrecorded: () => unrecorded,
    left: () => results.length - used.size + recall.length - recalled.size,
  };
}
